### PR TITLE
Initial support for running tests against WebKitTestRunner

### PR DESCRIPTION
**Pull request by gsnedders**
Obviously this shouldn't actually live in `executorcontentshell.py`, as there is a fair bit of divergence. It does seem like the content_shell code has problems with leaking files (pipes and semaphores, I think?), as a relatively modest number of restarts will cause macOS to run out of file handles for the process.

But, as evidence this mostly works:

```
gsnedders@gsnedders-marsha web-platform-tests % ./wpt run --no-restart-on-unexpected --no-manifest-update --log-mach - --log-mach-level info --binary /Users/gsnedders/projects/Safari/OpenSource/WebKitBuild/Release/WebKitTestRunner --manifest MANIFEST.json --metadata infrastructure/metadata wktr infrastructure
...
 1:11.56 INFO No more tests
 1:11.56 INFO Closing logging queue
 1:11.56 INFO queue closed
 1:11.59 INFO Got 5 unexpected results, with 1 unexpected passes
 1:11.59 SUITE_END

web-platform-test
~~~~~~~~~~~~~~~~~
Ran 297 checks (154 subtests, 143 tests)
Expected results: 253
Skipped: 39 tests
Unexpected results: 5
  test: 4 (1 fail, 1 ok, 2 timeout)
  subtest: 1 (1 fail)

Unexpected Results
------------------
/infrastructure/expected-fail/unhandled-rejection-following-subtest.html
  UNEXPECTED-OK /infrastructure/expected-fail/unhandled-rejection-following-subtest.html
/infrastructure/server/http2-websocket.sub.h2.any.worker.html
  TIMEOUT /infrastructure/server/http2-websocket.sub.h2.any.worker.html -
/infrastructure/server/http2-websocket.sub.h2.any.html
  TIMEOUT /infrastructure/server/http2-websocket.sub.h2.any.html -
/infrastructure/server/test-pac.html
  FAIL test that PAC metadata is respected - promise_test: Unhandled rejection with value: object "TypeError: Load failed"
get_stack@http://web-platform.test:8000/resources/testharness.js:4516:30
AssertionError@http://web-platform.test:8000/resources/testharness.js:4509:31
assert@http://web-platform.test:8000/resources/testharness.js:4493:37
@http://web-platform.test:8000/resources/testharness.js:759:35
@http://web-platform.test:8000/resources/testharness.js:2591:30
@http://web-platform.test:8000/resources/testharness.js:2638:40
/infrastructure/assumptions/ahem.html
  FAIL /infrastructure/assumptions/ahem.html - /infrastructure/assumptions/ahem.html ['242cbe5affeea2dfb65e72a2a717a4569cdcd018']
/infrastructure/assumptions/ahem-ref.html ['2de41e36cbb8b36868361c3170389cb84fc46ef9']
```

### CODE REVIEW

**Comment by gsnedders** [comment]
FWIW, I've been having problems with this failing with `OSError: [Errno 24] Too many open files` after n restarts. Adding a `gc.collect()` into `WKTRBrowser.stop` helps, but it looks like the circular reference that is preventing the `Process` object from being GC'd lives within `mozprocess`:

reference backgraph

cc @jgraham

File: tools/wpt/run.py [hunk 1]
```diff
@@ -686,6 +686,22 @@ def setup_kwargs(self, kwargs):
...
+
+    def install(self, channel=None):
+        raise NotImplementedError
+
+    def setup_kwargs(self, kwargs):
+        if kwargs["binary"] is None:
+            binary = self.browser.find_binary(self.venv.path, None)
+
+            if binary is None:
+                raise WptrunError("Unable to find servo binary in PATH")
```
Comment (jgraham) [inline tools/wpt/run.py:701]: Servo?

File: tools/wptrunner/wptrunner/browsers/wktr.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        self._extra_env = env
+
+    def start(self, group_metadata, **kwargs):
+        self.logger.debug("Starting WebKitTestRunner: %s..." % self._args[0])
+        #self.logger.debug(repr(self._args))
+
+        env = os.environ
+        env.update(self._extra_env)
+
+        # Unfortunately we need to use the Process class directly because we do not
```
Comment (jgraham) [inline tools/wptrunner/wptrunner/browsers/wktr.py:128]: Is there a reason not to just use `subprocess` in that case?

File: tools/wptrunner/wptrunner/browsers/wktr.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        underlying pipes have reached EOF.
+        """
+        return ExecutorBrowser, {"stdout_queue": self._stdout_queue,
+                                 "stderr_queue": self._stderr_queue,
+                                 "stdin_queue": self._stdin_queue}
+
+    def check_crash(self, process, test):
+        return not self.is_alive()
+
+    def _create_reader_thread(self, stream, queue):
```
Comment (jgraham) [inline tools/wptrunner/wptrunner/browsers/wktr.py:205]: Although I am mildly surprised that you can't use mozprocess's output handling, which iirc is pretty similar to what you have here.

File: tools/wptrunner/wptrunner/executors/executorwktr.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+        return result
+
+
+class WKTRProtocol(Protocol):
+    implements = [WKTRTestPart, WKTRErrorsPart]
+    init_timeout = 10  # Timeout (seconds) to wait for #READY message.
+
+    def connect(self):
+        """Waits for wktr to emit its "#READY" message which signals that it is fully
```
Comment (jgraham) [inline tools/wptrunner/wptrunner/executors/executorwktr.py:155]: It is surprising to me that this function has a docstring claiming it does something, but no body.

File: tools/wptrunner/wptrunner/browsers/wktr.py (hunk 2 above)
Comment (gsnedders) [reply tools/wptrunner/wptrunner/browsers/wktr.py:128]: This is mostly copied from `content_shell`, honestly. I _assumed_ there was a reason!

File: tools/wptrunner/wptrunner/executors/executorwktr.py (hunk 4 above)
Comment (gsnedders) [reply tools/wptrunner/wptrunner/executors/executorwktr.py:155]: Yeah, I should probably go and fix all the comments and docstrings here :|

File: tools/wpt/run.py (hunk 1 above)
Comment (gsnedders) [reply tools/wpt/run.py]: Gah. There's not really any point in having that function at all. Removed.

**Review by jgraham — APPROVED** [review]

**Pull request merged**